plugin/bullets.vim

```diff
@@ -573,6 +596,113 @@ endfun
+    " Only change the bullet level if it's currently a bullet.
+    let l:curr_line = l:curr_bullet.starting_at_line_num
+    let l:closest_bullet = s:closest_bullet_types(l:curr_line - g:bullets_line_spacing, l:curr_indent)
+    let l:closest_bullet = s:resolve_bullet_type(l:closest_bullet)
+
+    if l:closest_bullet != {}
+      let l:islower = l:closest_bullet.bullet ==# tolower(l:closest_bullet.bullet)
+      let l:closest_type = l:islower ? l:closest_bullet.bullet_type :
+            \ toupper(l:closest_bullet.bullet_type)
+      if l:closest_bullet.bullet_type ==# 'std'
+        let l:closest_type = l:closest_type . l:closest_bullet.bullet
+      endif
+      let l:closest_index = index(g:bullets_outline_levels, l:closest_type)
+
+      if l:closest_index >= 0
+        let l:closest_indent = indent(l:closest_bullet.starting_at_line_num)
+
+        if (l:curr_indent == l:closest_indent)
+          let l:next_bullet = s:next_bullet_str(l:closest_bullet)
+          let l:next_bullet_str = s:pad_to_length(l:next_bullet, l:closest_bullet.bullet_length)
+                \ . l:curr_bullet.text_after_bullet
+
+        elseif l:closest_index + 1 < len(g:bullets_outline_levels) || l:curr_indent < l:closest_indent
+          let l:next_index = l:closest_index + 1
+          let l:next_type = g:bullets_outline_levels[l:next_index]
+          let l:next_islower = l:next_type ==# tolower(l:next_type)
+          let l:trailing_space = ' '
+
+          let l:curr_bullet.closure = l:closest_bullet.closure
+
+          if l:next_type ==? 'rom'
+            let l:next_num = s:arabic2roman(1, l:next_islower)
+          elseif l:next_type ==? 'abc'
+            let l:next_num = s:dec2abc(1, l:next_islower)
+          elseif l:next_type ==# 'num'
+            let l:next_num = '1'
+          else
+            " standard bullet; l:next_type contains the bullet symbol to use
+            let l:next_num = strpart(l:next_type, len(l:next_type) - 1)
+            let l:curr_bullet.closure = ''
+          endif
+
+          let l:next_bullet_str = 
+                \ l:curr_bullet.leading_space
+                \ . l:next_num
+                \ . l:curr_bullet.closure
+                \ . l:trailing_space
+                \ . l:curr_bullet.text_after_bullet
+
+        else
+          " We're outside of the defined outline levels
+          let l:next_bullet_str =
+                \ l:curr_bullet.leading_space
+                \ . l:curr_bullet.text_after_bullet
+        endif
+
+        if l:next_bullet_str !=# ''
+          call setline(l:lnum, l:next_bullet_str)
+          execute "normal! $" 
+
+        else
+          if g:bullets_delete_last_bullet_if_empty
+            let l:orig_line = s:parse_bullet(l:lnum, getline(l:lnum))
+            if l:orig_line != []
+              call setline(l:lnum, l:orig_line[0].leading_space . l:orig_line[0].text_after_bullet)
+            endif
+          endif
+        endif
+      endif
+    endif
+  endif
```

Comment:
I'm a little concerned by all the `if` statements. Maybe there's a way to break some stuff into smaller private functions? Perhaps annotating with comments above each case what it is supposed to handle would help? (there's already some comments which I find very helpful but there are a lot of cases that don't)

Reply:
Yeah, I was looking for a cleaner way of handling this mess when I was writing it. For clarity's sake, I moved some of those checks into empty returns so that everything below it is more readable. I can offload some of this into other functions, but they would be single-use functions that require a bunch of parameters and this doesn't seem like the cleanest solution either.